content/blogposts/2022/multi-version-upgrades.md

```diff
@@ -0,0 +1,109 @@
+
+[Multi-version upgrades](https://docs.sourcegraph.com/admin/updates#multi-version-upgrades), contrasting to the "standard upgrade" approach, use the [`migrator` tool](https://about.sourcegraph.com/blog/introducing-migrator-service) to perform the sequence of schema and in-place data migrations that _would happen_ if the user were to perform a chain of upgrades that hit every minor release on the way to their target. This tool makes upgrading directly to our [4.0 release](https://about.sourcegraph.com/blog/release/4.0) a technical feasibility for the vast majority of our customers.
+
+At first glance, it may seem that multi-version upgrades are simply "standard upgrades in a `for`-loop". While that _is_ the basic idea, there's also a world of nuance surrounding it. Let's dive in.
+
+## The migration journey
+
+At the start of this journey, we used [golang-migrate](https://github.com/golang-migrate/migrate) to handle the details of our migrations. This tool would detect the current "schema state" of the database (stored in a forcibly-created table with at most one row) and apply any migrations that were defined strictly after the currently reported state. This process would happen on application startup, generally no-oping after the first run that actually did apply any missing migrations.
+
+Each migration exists on-disk as a pair of SQL files: one for the upgrade direction, and one for the downgrade direction. Each migration is defined in the same directory and ordered absolutely by a sequential numeric prefix. Unit tests ensure that the sequence has no gaps and that no migrations define the same prefix. The order of migration application is thereby determined by lexicographical ordering of the migration filenames.
+
+<Figure src="https://storage.googleapis.com/sourcegraph-assets/blog/multi-version-upgrades/treeview.png" alt="List of the (old) migrations directory." />
+
+This process didn't cause any clearly attributable problems for years. Around the time we started a serious hunt-to-kill effort for flaky tests in our CI pipelines, we started also [seeing initialization timeouts](https://github.com/sourcegraph/sourcegraph/issues/5721#issuecomment-540855232) we could attribute to reading and applying every single migration we've ever defined back-to-back.
+
+Because this problem would only get worse with time, we began to squash migration definitions by replacing the set of definitions at the head of the sequence with a single (yet equivalent) migration file. For Sourcegraph v3.9.0, we [squashed definitions](https://github.com/sourcegraph/sourcegraph/pull/5967) that were also defined in Sourcegraph v3.7.0 into to a single file. This cut startup time with an empty database from 20s to 5s, and also reduced the chance of hitting the startup timeout in our testing environment.
```

Comment:
```suggestion
Because this problem would only get worse with time, we began to squash migration definitions by replacing the set of definitions at the head of the sequence with a single (yet equivalent) migration file. For Sourcegraph v3.9.0, we [squashed definitions](https://github.com/sourcegraph/sourcegraph/pull/5967) that were also defined in Sourcegraph v3.7.0 into a single file. This cut startup time with an empty database from 20s to 5s and also reduced the chance of hitting the startup timeout in our testing environment.
```